### DISCUSSION TITLE
Command not found

### DISCUSSION THREAD

**Discussion by ju5t** [spatie/browsershot#580] [Q&A] [answered]
Last week I switched to a new laptop. The OS is MacOS (M1). When I try to create a PDF it gives the following error:

```bash
sh: npm: command not found sh: node: command not found
```

This is for the following command:

```
The command "PATH=$PATH:/usr/local/bin NODE_PATH=`npm root -g` node '/myhomefolder/vendor/spatie/browsershot/src/../bin/browser.js' '{"url":"file:\/\/\/var\/tmp\/486982453-0399117001637137714\/index.html","action":"pdf","options":{"args":[],"viewport":{"width":800,"height":600},"displayHeaderFooter":false,"format":"A4","margin":{"top":"60mm","right":"15mm","bottom":"40mm","left":"15mm"},"waitUntil":"networkidle0","printBackground":true}}'"
````

The error is pretty clear. It can't find the two commands. This feels like a `$PATH` issue. I tested the command as my own user and that works. The `$PATH` in PHP/Nginx is different though. It doesn't include `/opt/homebrew/bin/` (where these binaries live) by default.

I've added `/opt/homebrew/bin/` in `/etc/paths.d/homebrew` and restarted Valet. After the change, php.ini does give a `$PATH` which now includes these two paths. But the command still fails.

PHP is not restricted and should be able to access that path. When I run a basic selfmade script, it does work:

```php
<?php

$output = null;
exec('node --version', $output);
print_r($output);
```

So, eh, help? :)

edit: it does work when I use `setIncludePath`, but I prefer a server based solution instead. I prefer not to change the path in PHP.

**Comment by aerni** [comment, marked as answer]
I just created a PR that fixes this issue: https://github.com/spatie/browsershot/pull/581

**Reply by ju5t** [reply]
I completely missed this. Thanks for the PR!

**Reply by bryanjamesmiller** [reply]
I'm having a similar issue where the path on my computer that I need to use is: `/Users/bryanmiller/.nvm/versions/node/v14.17.6/bin` - but other developers/servers will have different paths and setups.  Any suggestions?

**Reply by plweil** [reply]
@bryanjamesmiller Did you find a fix for this? I'm having the exact same issue.

**Reply by plweil** [reply]
@bryanjamesmiller and others who might run into this. I solved this by setting the node and npm binaries:

```php
      $pdfContent = Browsershot::html($html)
        ->setNodeBinary('/Users/myhomedir/.nvm/versions/node/v14.18.2/bin/node')
        ->setNpmBinary('/Users/myhomedir/.nvm/versions/node/v14.18.2/bin/npm')
        ->pdf();
```

I'm not sure yet how to best manage this when node versions change.

**Reply by henrikruscon** [reply]
This was an issue for me as well, thanks @plweil.

Albeit not ideal for server environment, is nvm the issue here?

**Reply by plweil** [reply]
Yeah, having changeable versions of node, whether via nvm or whatever else, makes this a bit of a challenge to manage.

**Comment by joshuadegier** [comment]
Just adding my two cents to this topic. As I'm using node version manager the default paths still won't do the trick for my setup. As far as I can tell the only way to do this is using `setNodeBinary` and `setNpmBinary` which both would cause errors in production environments. Which could be worked around by using an environment setting holding the paths.

I suppose the best option you have if you run a similar setup as me is using `setIncludePath` where you would add your default path to the list as wel. For example:

`->setIncludePath('$PATH:/Users/macbook/.nvm/versions/node/v12.22.10/bin')`

Your path can be found by running `which npm` and `which node` in your terminal.

Still open for other suggestions, but I guess node version manager just is one of those things you really can't take into account in this package as the paths are different for everybody.

**Reply by marcorieser** [reply]
Same here

**Comment by lance2k** [comment]
This was fixed with setIncludePath, but other developers/servers will have different paths and setups. Any suggestions other than reverting to nodesource installer?

**Comment by benlumley** [comment]
To to make this nicer.... I've gone with this in AppServiceProvider
```
    $this->app->extend(PdfFactory::class, function (PdfFactory $service)  : PdfFactory {
            $service->default()->withBrowsershot(function (Browsershot $browsershot): void {
                // this is to make nvm node be used - only needed on local
                if (!empty(config('app.node_path')) && is_dir(config('app.node_path'))) {
                    $browsershot->setIncludePath('$PATH:' . config('app.node_path'));
                }
            });
            return $service;
        }); 
```

+ in config/app.php - 'node_path' => env('NODE_PATH'),

Then you set the NODE_PATH env var if you need it; if not set - gets ignored. 
Makes this hack workable for me.

**Reply by benlumley** [reply]
oh... wrong repo; kinda. I'm using spatie/laravelpdf
